File: Golang/653. Two Sum IV - Input is a BST.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
// func findTarget(root *TreeNode, k int) bool { // 深度搜索 44ms
//     node_list := make([]TreeNode, 0)
//     node_list = append(node_list, *root)
//     nums := make(map[int]bool, 0)
    
//     for len(node_list) > 0 {
//         node := node_list[len(node_list)-1]
//         node_list = node_list[:len(node_list)-1]
//         if nums[k - node.Val] {
//             return true
//         } else {
//             nums[node.Val] = true
//             if node.Left != nil {
//                 node_list = append(node_list, *node.Left)
//             }
//             if node.Right != nil {
//                 node_list = append(node_list, *node.Right)
//             }
//         }
//     }
    
//     return false
// }

func findTarget(root *TreeNode, k int) bool { // 广度搜索 56ms
    cur := []*TreeNode{root}
    nextlevel := make([]*TreeNode, 0)
    nums := make(map[int]bool, 0)
    
    for len(cur) > 0 {
        for _, node := range cur {
            if nums[k - node.Val] {
                return true
            } else {
                nums[node.Val] = true
                if node.Left != nil {
                    nextlevel = append(nextlevel, node.Left)
                }
                if node.Right != nil {
                    nextlevel = append(nextlevel, node.Right)
                }
            }
        }
        cur = nextlevel
        nextlevel = []*TreeNode{}
    }
    
    return false
}

File: Python3/653. Two Sum IV - Input is a BST.py
class Solution:
    def findTarget(self, root, k):
        """
        :type root: TreeNode
        :type k: int
        :rtype: bool
        """
        # node_list = [root] # 深度搜索 # 112ms
        # nums = set()      
        # while node_list:
        #     node = node_list.pop()
        #     if k - node.val in nums:
        #         return True
        #     nums.add(node.val)
        #     if node.left: node_list.append(node.left)
        #     if node.right: node_list.append(node.right)
        # return False
        
        cur = [root] # 广度搜索 # 104ms
        nextlevel = []
        nums = set()
        while cur:
            for node in cur:
                if k - node.val in nums:
                    return True
                nums.add(node.val)
                if node.left: nextlevel.append(node.left)
                if node.right: nextlevel.append(node.right)
            cur = nextlevel
            nextlevel = []
        return False
